### PR TITLE
SNOW-664933 Fix write_pandas with empty dataframe

### PR DESCRIPTION
**Pull request by sfc-gh-stan**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #SNOW-664933

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am modifying authorization mechanisms
   - [ ] I am adding new credentials
   - [ ] I am modifying OCSP code
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

  `write_pandas` used to raise `ValueError` when we attempt to write a dataframe with no rows, this is caused by `chunk_helper` attempting to generate chunks of 0 rows (`for i in range(0, len(lst), n):` when `n=0`). With this code change, we will just generate one chunk when the dataframe has fewer rows than the chunk size, which includes the special case of 0 rows.

### CODE REVIEW

File: test/integ/pandas/test_pandas_tools.py [hunk 1]
```diff
@@ -300,6 +300,21 @@ def test_invalid_table_type_write_pandas(
             )
 
 
+def test_empty_dataframe_auto_create_table_write_pandas(
```
Comment (sfc-gh-stan) [inline test/integ/pandas/test_pandas_tools.py:303]: The test name seems redundant, the issue we fix is irrelevant to `auto_create_table`.

**Review by sfc-gh-sfan — APPROVED** [review]
LGTM. Can you also port this to sp connector?

File: src/snowflake/connector/pandas_tools.py [hunk 2]
```diff
@@ -35,6 +35,9 @@
 
 def chunk_helper(lst: T, n: int) -> Iterator[tuple[int, T]]:
     """Helper generator to chunk a sequence efficiently with current index like if enumerate was called on sequence."""
+    if n >= len(lst):
+        yield 0, lst
+        return
```
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/pandas_tools.py:40]: So it worked for shorter lists already:
```
>>> list(chunk_helper([1, 2], 3))
[(0, [1, 2])]
```
There's only one case the original function didn't work for and so this is more explicit:
```suggestion
    if n == len(lst):
        yield 0, lst
        return
```

**Review by sfc-gh-mkeller — CHANGES_REQUESTED** [review]

File: test/integ/pandas/test_pandas_tools.py [hunk 3]
```diff
@@ -300,6 +300,21 @@ def test_invalid_table_type_write_pandas(
...
+    table_name = random_string(5, "empty_dataframe_")
+    df = pandas.DataFrame([], columns=["name", "balance"])
+    with conn_cnx() as cnx:
+        try:
+            success, num_chunks, num_rows, _ = write_pandas(
+                cnx, df, table_name, auto_create_table=True
+            )
+            assert success and num_chunks == 1 and num_rows == 0
+        finally:
+            cnx.execute_string(f"DROP TABLE IF EXISTS {table_name}")
```
Comment (sfc-gh-mkeller) [inline test/integ/pandas/test_pandas_tools.py:315]: No need to call that helper fn (execute is enough), also if we used a temp table then this wouldn't be necessary at all.
```suggestion
            cnx.execute(f"DROP TABLE IF EXISTS {table_name}")
```

File: test/integ/pandas/test_pandas_tools.py [hunk 4]
```diff
@@ -300,6 +300,21 @@ def test_invalid_table_type_write_pandas(
...
 
+def test_empty_dataframe_write_pandas(
+    conn_cnx: Callable[..., Generator[SnowflakeConnection, None, None]],
+):
+    table_name = random_string(5, "empty_dataframe_")
+    df = pandas.DataFrame([], columns=["name", "balance"])
+    with conn_cnx() as cnx:
+        try:
+            success, num_chunks, num_rows, _ = write_pandas(
+                cnx, df, table_name, auto_create_table=True
```
Comment (sfc-gh-mkeller) [inline test/integ/pandas/test_pandas_tools.py:311]: We could use a temp table for this:
```suggestion
                cnx, df, table_name, auto_create_table=True, table_type="temp",
```

**Review by sfc-gh-mkeller — APPROVED** [review]
:ship: :ship:

**Pull request merged**